File: Objective-C/Shields/Shields/ShieldsLabel.m
//
//  ShieldsLabel.m
//  Shields
//
//  Created by Mac on 2021/5/21.
//

#import "ShieldsLabel.h"

@interface ShieldsLabel()
@property (nonatomic,strong)UIImageView *logo;
@property (nonatomic,strong)UILabel *label;
@end

@implementation ShieldsLabel

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        /// 配置UI
        [self setupUI:ShieldsDefault];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame style:(ShieldsStyle)style {
    self = [super initWithFrame:frame];
    if (self) {
        /// 配置UI
        [self setupUI:style];
    }
    return self;
}

#pragma mark - 配置UI
- (void)setupUI:(ShieldsStyle)style {
    self.backgroundColor = [UIColor colorWithRed: 93/255.0 green: 93/255.0 blue: 93/255.0 alpha:1.000];
    
    if (style == ShieldsDefault) {
        [self onlyLabel];
    } else if (style == ShieldsLogo) {
        [self logoWithLabel];
    }
}

- (void)onlyLabel {
    CGFloat paddingL = 6;
    CGFloat paddingR = 4;
    CGFloat paddingY = 4;
    
    CGFloat labelX = paddingL;
    CGFloat labelY = paddingY;
    CGFloat labelW = CGRectGetWidth(self.bounds) - paddingL - paddingR;
    CGFloat labelH = CGRectGetHeight(self.bounds) - paddingY * 2.0;
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(labelX, labelY, labelW, labelH)];
    label.textColor = UIColor.whiteColor;
    label.text = @"label";
    label.textAlignment = NSTextAlignmentCenter;
    [self addSubview:label];
    self.label = label;
}

- (void)logoWithLabel {
    /// logo
    CGFloat logoX = 2;
    CGFloat logoY = 4;
    CGFloat logoW = 20;
    CGFloat logoH = logoW;
    UIImageView *logo = [[UIImageView alloc] initWithFrame: CGRectMake(logoX, logoY, logoW, logoH)];
    logo.center = CGPointMake(logoX+logoW*0.5+5, self.center.y);
    NSString *errorLogoPath = [[NSBundle mainBundle] pathForResource:@"errorLogo" ofType:@"png"];
    logo.image = [[UIImage alloc] initWithContentsOfFile:errorLogoPath];
    logo.image = [logo.image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    [logo setTintColor:UIColor.whiteColor];
    [self addSubview:logo];
    self.logo = logo;
    
    /// label
    CGFloat labelX = logoX + logoW + 2;
    CGFloat labelY = logoY;
    CGFloat labelW = CGRectGetWidth(self.bounds) - labelX - 2;
    CGFloat labelH = CGRectGetHeight(self.bounds) - logoY * 2.0;
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(labelX, labelY, labelW, labelH)];
    label.textColor = UIColor.whiteColor;
    label.text = @"label";
    label.textAlignment = NSTextAlignmentCenter;
    [self addSubview:label];
    self.label = label;
}

#pragma mark - 基本配置
- (void)configLabel:(NSString *)str {
    self.label.text = str;
}

- (void)configLabel:(NSString *)str color:(UIColor *)color {
    [self configLabel:str];
    self.label.textColor = color;
}

- (void)configLogo:(UIImage *)logo {
    self.logo.image = logo;
}

- (void)configLogo:(UIImage *)logo logoWidth:(CGFloat)logoWidth {
    [self configLogo:logo];
    
    if (logoWidth != 0.0) {
        self.logo.frame = CGRectMake(self.logo.frame.origin.x, self.logo.frame.origin.y, logoWidth, logoWidth);
    }
}

- (void)configLogo:(UIImage *)logo logoWidth:(CGFloat)logoWidth logoPosition:(CGPoint)logoPosition {
    [self configLogo:logo logoWidth:logoWidth];
    self.logo.center = logoPosition;
}

@end

File: Swift/Shields/Shields/Shields/ShieldsLabel.swift
//
//  ShieldsLabel.swift
//  Shields
//
//  Created by 纽扣软件 on 2021/5/24.
//

import UIKit

class ShieldsLabel: UIView {
    enum ShieldsStyle: Int {
        case ShieldsDefault
        case ShieldsLogo
    }
    
    private var logo: UIImageView = UIImageView()
    private var label: UILabel = UILabel()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        /// 配置UI
        setupUI(style: .ShieldsDefault)
    }
    
    init(frame: CGRect, style: ShieldsStyle) {
        super.init(frame: frame)
        
        /// 配置UI
        setupUI(style: style)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

//MARK:- 配置UI
extension ShieldsLabel {
    private func setupUI(style: ShieldsStyle) {
        backgroundColor = UIColor(red: 93/255.0, green: 93/255.0, blue: 93/255.0, alpha: 1.0)
        
        switch style {
        case .ShieldsDefault:
            onlyLabel()
        case .ShieldsLogo:
            logoWithLabel()
        }
    }
    
    private func onlyLabel() {
        let paddingL: CGFloat = 6
        let paddingR: CGFloat = 4
        let paddingY: CGFloat = 4
        
        let labelX: CGFloat = paddingL
        let labelY: CGFloat = paddingY
        let labelW: CGFloat = self.bounds.width - paddingL - paddingR;
        let labelH: CGFloat = self.bounds.height - paddingY * 2.0;
        let label: UILabel = UILabel(frame: CGRect(x: labelX, y: labelY, width: labelW, height: labelH))
        label.textColor = .white
        label.text = "label"
        label.textAlignment = .center
        addSubview(label)
        self.label = label;
    }
    
    private func logoWithLabel() {
        /// logo
        let logoX: CGFloat = 2
        let logoY: CGFloat = 4
        let logoW: CGFloat = 20
        let logoH: CGFloat = logoW
        
        let logo: UIImageView = UIImageView(frame: CGRect(x: logoX, y: logoY, width: logoW, height: logoH))
        logo.center = CGPoint(x: logoX+logoW*0.5+5, y: self.center.y);
        if let errorLogoPath: String = Bundle.main.path(forResource:"errorLogo", ofType: "png") {
            logo.image = UIImage(contentsOfFile: errorLogoPath)
        }
        
        addSubview(logo)
        self.logo = logo
        
        /// label
        let labelX: CGFloat = logoX + logoW + 2
        let labelY: CGFloat = logoY
        let labelW: CGFloat = self.bounds.width - labelX - 2
        let labelH: CGFloat = self.bounds.height - logoY * 2.0
        let label: UILabel = UILabel(frame: CGRect(x: labelX, y: labelY, width: labelW, height: labelH))
        label.textColor = UIColor.white
        label.text = "label"
        label.textAlignment = .center;
        addSubview(label)
        self.label = label;
    }
}

//MARK:- 基本配置
extension ShieldsLabel {
    func configLabel(str: String) {
        label.text = str
    }
    
    func configLabel(str: String, color: UIColor) {
        configLabel(str: str)
        label.textColor = color
    }
    
    func configLogo(img: UIImage) {
        logo.image = img
    }
    
    func configLogo(img: UIImage, logoWidth: CGFloat) {
        configLogo(img: img)
        
        if logoWidth != 0.0 {
            logo.frame = CGRect(x: self.logo.frame.origin.x, y: self.logo.frame.origin.y, width: logoWidth, height: logoWidth)
        }
    }
    
    func configLogo(img: UIImage, logoWidth: CGFloat, logoPosition: CGPoint) {
        configLogo(img: img, logoWidth: logoWidth)
        logo.center = logoPosition
    }
}
